scheduling/storage_providers/memory.go

```diff
@@ -2,48 +2,186 @@ package scheduling
 
 import (
 	"fmt"
+	"sort"
 	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type LockInformation struct {
+	ID   string
+	Key  string
+	Date time.Time
+	Lock LockObject
+}
+
+const (
+	UpdateSleepTime = 2 * time.Second
+	ValidTimePeriod = 5 * time.Second
+	CleanupInterval = 10 * time.Second
 )
 
 type MemoryStorageProvider struct {
-	storage map[string]string
+	storage     sync.Map
+	lockedItems sync.Map
 }
 
 func NewMemoryStorageProvider() *MemoryStorageProvider {
-	return &MemoryStorageProvider{storage: make(map[string]string)}
+	storage := sync.Map{}
+	lockedItems := sync.Map{}
+	return &MemoryStorageProvider{storage: storage, lockedItems: lockedItems}
 }
 
-func (m *MemoryStorageProvider) Set(key string, value string) error {
+func (m *MemoryStorageProvider) Set(key string, value string, lock LockObject) error {
+	lockInfo, ok := m.lockedItems.Load(key)
+	if !ok {
+		return fmt.Errorf("key is not locked")
+	}
+
+	currentLock := lockInfo.(LockInformation)
+	if currentLock.ID != lock.ID {
+		return fmt.Errorf("key is locked by another id: locked by: %s, unlock by: %s", currentLock.ID, lock.ID)
+	}
+
 	if key == "" {
 		return fmt.Errorf("key is empty")
 	}
 	if value == "" {
 		return fmt.Errorf("value is empty for key %s", key)
 	}
-	m.storage[key] = value
+	m.storage.Store(key, value)
 	return nil
 }
 
-func (m *MemoryStorageProvider) Get(key string, prefix bool) ([]string, error) {
+func (m *MemoryStorageProvider) Get(key string, prefix bool) (map[string]string, error) {
 	if key == "" {
 		return nil, fmt.Errorf("key is empty")
 	}
+
+	result := make(map[string]string)
+
 	if !prefix {
-		value, ok := m.storage[key]
+		value, ok := m.storage.Load(key)
 		if !ok {
 			return nil, &KeyNotFoundError{Key: key}
 		}
-		return []string{value}, nil
+		result[key] = value.(string)
+		return result, nil
 	}
 
-	var result []string
-	for k, v := range m.storage {
-		if strings.HasPrefix(k, key) {
-			result = append(result, v)
+	// loops through the keys in sync map
+	// and finds the keys that have the prefix
+	m.storage.Range(func(k, v interface{}) bool {
+		mapKey := k.(string)
+		if strings.HasPrefix(mapKey, key) {
+			result[mapKey] = v.(string)
 		}
-	}
+		return true
+	})
+
 	if len(result) == 0 {
 		return nil, &KeyNotFoundError{Key: key}
 	}
 	return result, nil
 }
+
+func (m *MemoryStorageProvider) ListKeys(prefix string) ([]string, error) {
+	var result []string
+	// loops through the keys in sync map
+	// and finds the keys that have the prefix
+	m.storage.Range(func(k, v interface{}) bool {
+		mapKey := k.(string)
+		if strings.HasPrefix(mapKey, prefix) {
+			result = append(result, mapKey)
+		}
+		return true
+	})
+	sort.Strings(result)
+
+	return result, nil
+}
+
+type LockObject struct {
+	ID      string
+	Channel *chan error
+}
+
+func (m *MemoryStorageProvider) Lock(key string) (LockObject, error) {
+	id := uuid.New().String()
```

Comment:
nit: you can save a few machine cycles by doing the id assignment until after the key == "" validation check.